src/model_theory/definability.lean

```diff
@@ -12,68 +12,163 @@ import model_theory.terms_and_formulas
 This file defines what it means for a set over a first-order structure to be definable.
 
 ## Main Definitions
-* A `first_order.language.definable_set` is defined so that `L.definable_set M α` is the boolean
-  algebra of subsets of `α → M` defined by formulas.
+* `first_order.language.is_definable` is defined so that `L.is_definable A s` indicates that the
+set `s` of a finite cartesian power of `M` is definable with parameters in `A`.
+* `first_order.language.is_definable₁` is defined so that `L.is_definable₁ A s` indicates that
+`(s : set M)` is definable with parameters in `A`.
+* `first_order.language.is_definable₂` is defined so that `L.is_definable₂ A s` indicates that
+`(s : set (M × M))` is definable with parameters in `A`.
+* A `first_order.language.definable_set` is defined so that `L.definable_set α A` is the boolean
+  algebra of subsets of `α → M` defined by formulas with parameters in `A`.
 
 ## Main Results
-* `L.definable_set M α` forms a `boolean_algebra.
+* `L.definable_set M α` forms a `boolean_algebra`.
+* `is_definable.image_comp_sum` shows that definability is closed under projections.
 
 -/
 
+universes u v w
+
 namespace first_order
 namespace language
 
-variables {L : language} {M : Type*} [L.Structure M]
+variables {L : language.{u v}} {M : Type w} [L.Structure M] {A : set M}
 open_locale first_order
-open Structure
+open Structure set
 
 /-! ### Definability -/
 
 section definability
 
-variables (L) {α : Type} [fintype α]
+variables (L) {α : Type} [fintype α] {β : Type} [fintype β] (A)
 
-/-- A subset of a finite Cartesian product of a structure is definable when membership in
-  the set is given by a first-order formula. -/
+/-- A subset of a finite Cartesian product of a structure is definable over a set `A` when
+  membership in the set is given by a first-order formula with parameters from `A`. -/
 structure is_definable (s : set (α → M)) : Prop :=
-(exists_formula : ∃ (φ : L.formula α), s = set_of φ.realize)
+(exists_formula : ∃ (φ : L[[A]].formula α), s = set_of φ.realize)
+
+variables {L} {A} {B : set M} {s : set (α → M)}
+
+lemma is_definable.map_expansion {L' : language} [L'.Structure M] (h : L.is_definable A s)
+  (φ : L →ᴸ L') [φ.is_expansion_on M] :
+  L'.is_definable A s :=
+begin
+  obtain ⟨ψ, rfl⟩ := h,
+  refine ⟨⟨(φ.add_constants A).on_formula ψ, _⟩⟩,
+  ext x,
+  simp only [mem_set_of_eq, Lhom.realize_on_formula],
+end
+
+lemma is_empty_definable_iff :
+  L.is_definable ∅ s ↔ ∃ (φ : L.formula α), s = set_of φ.realize :=
+begin
+  split,
+  { rintro ⟨φ, rfl⟩,
+    refine ⟨(L.Lhom_trim_empty_constants (∅ : set M)).on_formula φ, _⟩,
+    ext x,
+    simp only [mem_set_of_eq, Lhom.realize_on_formula], },
+  { rintro ⟨φ, rfl⟩,
+    refine ⟨⟨(L.Lhom_with_constants (∅ : set M)).on_formula φ, _⟩⟩,
+    ext x,
+    simp only [mem_set_of_eq, Lhom.realize_on_formula], }
+end
+
+lemma is_definable_iff_empty_definable_with_params :
```

Comment:
Here and in other places the names don't always follow a clear pattern: is it `is_definable` or `definable`? You also seem to use `(is_)definable` in an infix manner (which makes sense from an English pov). Is there something we can do to also reflect that in the Lean code?

Reply:
I suppose I should just use `definable`, unless I somehow use `s.is_definable L A` as my variable order.

I guess to match the infix style, I'd have to make it `set.definable`, where `A.definable L s` means that `s` is definable with parameters from `A`, as most closely matches that English usage of "s is A-definable".

So, which sounds better - `set.definable` and `A.definable L s`, or `language.definable` and `L.definable A s`?

(The `L` can't go away - there are definitely setups where you need to juggle different languages.)